Parse Designators

skas102 opened the pull request.

Commit:
Add member and element access to parser

- Create MemberAccessNode
- Create ElementAccessNode
- Modify Visitors
- Fix Test Utils
- Testing
- Refactoring

Commit:
Designators

- Added TypeSymbol Interface
- Added Array.length
- Refactoring

skas102 commented on parser/parser_expression.go:
```diff
@@ -168,16 +168,39 @@ func (p *Parser) parseOperand() node.ExpressionNode {
-func (p *Parser) parseDesignator() *node.DesignatorNode {
-	return &node.DesignatorNode{
+func (p *Parser) parseDesignator() node.Node {
+	var left node.Node
```
Not possible... If I would merge thos two, left would be of type DesignatorNode but is later values of type DesignatorNode, ElementAccessNode and MemberAccessNode are assigned to it.

tk-codes replied:
@rpfister102 Could be resolved like this https://github.com/bazo-blockchain/lazo/pull/55/commits/6b69b36e48cde169f015198f7321ed9ea519d08c

tk-codes commented on parser/node/nodes.go:
```diff
@@ -388,22 +388,58 @@ func (n *DesignatorNode) Accept(v Visitor) {
 type FuncCallNode struct {
 	AbstractNode
-	Designator *DesignatorNode
+	Designator Node
```
It is better to create DesignatorNode interface than using the Node interface.

tk-codes commented on parser/parser.go:
```diff
@@ -292,29 +292,29 @@ func (p *Parser) parseMultiVariableStatement(varType *node.TypeNode, id string)
 
 	return &node.AssignmentStatementNode{
-		AbstractNode: left.AbstractNode,
+		AbstractNode: p.newAbstractNode(),
```
Assignment's position should be the position of designator

tk-codes replied:
Fixed https://github.com/bazo-blockchain/lazo/pull/55/commits/695f45cfe8798a3abf2696edbdb9dcaa87988c35

tk-codes commented on parser/parser_expression_test.go:
```diff
@@ -254,6 +254,53 @@ func TestFuncCallMissingComma(t *testing.T) {
+func TestInvalidMemberAccess(t *testing.T) {
+	p := newParserFromInput("a.0")
+	p.parseExpression()
+	assertHasError(t, p)
```
Use `assertErrorAt` for checking errors

tk-codes replied:
Fixed https://github.com/bazo-blockchain/lazo/pull/55/commits/50bbce4be94f797c8c6332a68f58e75a84466e61

tk-codes committed:
Rename DesignatorNode to BasicDesignatorNode

tk-codes committed:
Rename VisitDesignatorNode to VisitBasicDesignatorNode

tk-codes committed:
Use DesignatorNode interface

Node interface is too generic. It is better to specify a separate concrete designator interface

tk-codes committed:
Merge variable declaration with assignment on next line

tk-codes committed:
Use pointers to basic type symbol

DO NOT copy basic type symbols. Use pointers to the original type instances (int, bool etc.)

tk-codes committed:
Fix designator node's position

Function call, member access designator, element access designator, assignments etc. have the position of their first designator

tk-codes committed:
Use DesignatorNode in MultiAssignment and CallStatement

tk-codes committed:
Fix symbol table functions to use designator interface

tk-codes committed:
Check for errors explicitely

tk-codes committed:
Remove unnecessary assertions

assertErrorAt also asserts if it has errors, therefore it is not necessary to assert twice

tk-codes reviewed (approved):
@rpfister102 I fixed the requested changes on my own in order to continue with other features
